### PR TITLE
Add ability to select / unselect repositories

### LINKED ISSUE #184: Ability to select analyzed repositories
### Story

When browsing a profile (organization or user), I want to be able to select/unselect analyzed repositories.

Currently, all repositories are analyzed and this should stay the default, all checked by default.

### Scenario

1. Go on https://backyourstack.com/
2. Go to an organization or user profile, ie: https://backyourstack.com/airbnb
3. Go to "Analyzed Repositories"
4. All repositories are checked by default. I'm able to uncheck/recheck them.
5. When switching back to "Projects requiring funding" or "Detected Dependencies", I can immediately see the result

### PR DESCRIPTION
**Pull request by flickz**
Resolves #184 

Screenshot 2019-10-17 at 9 37 12 AM

TODO: 

- [x]  Display only selected repository dependencies from in `monthly-plan` page.
- [x]  Add a way to remember filtered repositories.

### CODE REVIEW

File: src/components/RepositoryTable.js [hunk 1]
```diff
@@ -4,6 +4,7 @@ import PropTypes from 'prop-types';
 export default class RepositoryTable extends React.Component {
   static propTypes = {
     repositories: PropTypes.array.isRequired,
+    onSelectRepo: PropTypes.func,
```
Comment (znarf) [inline src/components/RepositoryTable.js]: should be `onSelectRepository`, we try to avoid jargon and the word `repo` in that case.

File: src/lib/data.js [hunk 2]
```diff
@@ -1,3 +1,4 @@
+/* eslint-disable require-atomic-updates */
```
Comment (znarf) [inline src/lib/data.js]: This should be in the global eslint configuration.

File: src/pages/files.js [hunk 3]
```diff
@@ -29,9 +24,10 @@ export default class Files extends React.Component {
 
     // sessionFiles is optional and can be null (always on the client)
     const sessionFiles = get(req, 'session.files');
-    const { files, dependencies, recommendations } = await getFilesData(
+    const { files, dependencies, recommendations } = await getData({
```
Comment (znarf) [inline src/pages/files.js]: I'm confused with this. What's the rationale behind this change?

File: src/server/index.js [hunk 4]
```diff
@@ -146,10 +146,17 @@ nextApp.prepare().then(() => {
   server.get('/data/getProfileData', (req, res) => {
     const accessToken = get(req, 'session.passport.user.accessToken');
     const loggedInUsername = get(req, 'session.passport.user.username');
+    const profileOpts = { loggedInUsername };
```
Comment (znarf) [inline src/server/index.js]: Should be `profileOptions`

**Comment by znarf** [comment]
Can we find a way to make the "Back My Stack" -> "Continue on Open Collective" link works on Now automated deployments?

I don't think having excludedRepos in the URL is the best solution. Have you considered alternatives?

File: src/pages/files.js (hunk 3 above)
Comment (flickz) [reply src/pages/files.js]: It was just a small abstraction, I had `getFilesData` and `getProfileData` on several pages, I needed to modify their arguments, I find myself updating them in multiple files so I decided bring it all together as a single method in a single file.

**Comment by flickz** [comment]
> Can we find a way to make the "Back My Stack" -> "Continue on Open Collective" link works on Now automated deployments?

I'll see to that.

> I don't think having excludedRepos in the URL is the best solution. Have you considered alternatives?

It's about persisting `excludedRepos` in case of page reload and navigating to other pages, without losing them. I had another options of saving them in browser local storage but I felt having them in url provides more flexibility since you can modify them directly there without going to `Analyzed Repositories` tab. 


> Overall, the code base starts to be more and more complicated to navigate, what can we do to make it simpler, easier to understand?

That's true, I'll need time to look into this.

File: src/pages/profile.js [hunk 5]
```diff
@@ -128,7 +128,7 @@ export default class Profile extends React.Component {
     }
   };
 
-  hanldeOnSelectRepo = ({ target }) => {
+  hanldeOnSelectRepository = ({ target }) => {
```
Comment (znarf) [inline src/pages/profile.js]: typo -> `handleOnSelectRepository`

**Comment by flickz** [comment]
Update here:

> I'm confused about the getData abstraction, It's clearer to me without it. Let's revert that.

Reverted 

> Let's not put excludedRepos in the backing.json URL. It should just stay https://backyourstack-keynultlmp.now.sh/znarf/profile/backing.json. Whatever is required should be written in the S3 bucket.

Taken care of.

> Let's try to have the excludedRepos state elsewhere than in the URL.

This is difficult, as the data are needed in monthly-plan page,  query string seems like the easiest and simplest way to pass the data, it also has the advantage of quickly excluding repositories from the url instead of going through a  long list of checkboxes so I'm leaving it.

**Comment by znarf** [comment]
@flickz the PR looks messed up with "Update confirmation page for asynchronous dispatch". https://github.com/opencollective/backyourstack/pull/250

It's complicated to review in this state.

**Comment by znarf** [comment]
@flickz it seems the version is now messed up, I see a lot of things that are supposed to be addressed (getData, eslint)

**Comment by flickz** [comment]
I apologized I've reset the HEAD commit back to what it should be.

File: src/server/index.js [hunk 6]
```diff
@@ -216,36 +223,50 @@ nextApp.prepare().then(() => {
...
     return res.status(200).send({ id: savedId });
   });
 
   server.get('/:id/backing.json', async (req, res) => {
     const accessToken = get(req, 'session.passport.user.accessToken');
-    const profileData = await getProfileData(req.params.id, accessToken);
+    let excludedRepos;
+    if (req.query.excludedRepos) {
+      excludedRepos = JSON.parse(req.query.excludedRepos);
+    }
```
[the remark this answers is not in the document]
Comment (flickz) [reply src/server/index.js]: Yeah I did implement the feedback for `backing.json` endpoint we use for dispatch, relying on s3. For this endpoint, we use it to display data in `profile.json` page, it is really important to filter repository from  [`getProfileData`](https://github.com/opencollective/backyourstack/blob/feat/select-repos/src/lib/data.js#L45).

File: src/pages/profile.js [hunk 7]
```diff
@@ -6,7 +6,7 @@ import { get } from 'lodash';
 
 import { Link, Router } from '../routes';
 
-import { fetchJson, postJson } from '../lib/fetch';
```
[the remark this answers is not in the document]
Comment (flickz) [reply src/pages/profile.js]: I'm sorry, I already messed up the commits, I had to manually update that back, the mistake was `postJson` coming before `fetchJson` but I don't think that matters technically.

File: src/server/index.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (znarf) [reply src/server/index.js]: Sorry, I don't get why we have `excludedRepos` here if `backing.json` is operating without parameters.

[the remark this answers is not in the document]
Comment (flickz) [reply src/server/index.js]: I'm passing it as query string not a route parameter path.

[the remark this answers is not in the document]
Comment (znarf) [reply src/server/index.js]: Ok, my initial feedback was more about query string parameters than route parameters.

I still don't see the purpose of `excludedRepos` here.

[the remark this answers is not in the document]
Comment (flickz) [reply src/server/index.js]: The purpose is to be able to use it in [`getProfileData`](https://github.com/opencollective/backyourstack/blob/feat/select-repos/src/lib/data.js#L45) if you look [here](https://github.com/opencollective/backyourstack/blob/feat/select-repos/src/pages/profile.js#L31-L35)  in profile.js`excludedRepos` is passed to `getProfileData` directly for client, we want to be able to do the same [here](https://github.com/opencollective/backyourstack/blob/feat%2Fselect-repos/src/server/index.js#L159-L161) from api hence why it is passed as a query string.

[the remark this answers is not in the document]
Comment (znarf) [reply src/server/index.js]: - `backing.json` should not have query string parameters, the stored profile on S3 should be enough. I don't get the use case.

**Comment by znarf** [comment]
Looks like there is 3 different endpoints for `backing.json`, I thought we only had 2 different use cases. Can you detail what are now the use cases for each of these endpoints?

**Comment by znarf** [comment]
When Backing My Stack using the feature, it's crashing.

Test URL: https://backyourstack-jbcfetvkqg.now.sh/monthly-plan?id=znarf&type=profile&excludedRepos=advent-of-code-2018%2Camateur%2Cbbpress-minimal%2Cblogmarks3%2Cbouncer%2Cdokuwiki-css%2Cdokuwiki-minimal%2Cdokuwiki-openid%2Ch6e-minimal%2Cnvo%2Copencollective-exercise%2Cresume%2Cstatusnet-ladistribution%2Cwordpress-minimal

Screenshot 2019-11-11 at 13 02 50

File: src/lib/data.js (hunk 2 above)
Comment (znarf) [reply src/lib/data.js]: Why was that resolved?

File: src/lib/data.js [hunk 8]
```diff
@@ -37,20 +38,32 @@ export async function getProfileData(id, accessToken, loggedInUsername) {
...
+            repo.checked = false;
+            repo.dependencies = [];
+          }
+        } else {
+          repo.dependencies = await getDependenciesFromGithubRepo(
+            repo,
+            accessToken,
+          );
+          repo.checked = true;
+        }
```
Comment (znarf) [inline src/lib/data.js]: This block can be simplified so we don't have the duplicated code section and nested if/else.

File: src/lib/fetch.js [hunk 9]
```diff
@@ -21,3 +34,13 @@ export function postJson(url, body, params = {}) {
   params.body = JSON.stringify(body);
   return fetchJson(url, params);
 }
+
+export function getData(options = {}) {
```
Comment (znarf) [inline src/lib/fetch.js]: This abstraction doesn't make things easier to follow. I think `getFilesData` can be interesting.

The switch between `getProfileData` and `getFilesData` can simply happen where it's needed.

File: src/pages/profile.js [hunk 10]
```diff
@@ -23,10 +23,18 @@ import FacebookLogo from '../static/img/facebook.svg';
 
 const ocWebsiteUrl = process.env.WEBSITE_URL || 'https://opencollective.com';
 
-const getProfileData = (id, accessToken) =>
-  process.env.IS_CLIENT
-    ? fetchJson(`/data/getProfileData?id=${id}`)
-    : import('../lib/data').then(m => m.getProfileData(id, accessToken));
+const getProfileData = (id, accessToken, excludedRepos) => {
```
Comment (znarf) [inline src/pages/profile.js]: Why not have `(id, accessToken, { excludedRepos } )` instead of `(id, accessToken, excludedRepos)`

File: src/pages/profile.js (hunk 5 above)
Comment (znarf) [reply src/pages/profile.js]: Why was that resolved?

File: src/server/index.js (hunk 4 above)
Comment (znarf) [reply src/server/index.js]: Why was that resolved?

File: src/server/index.js [hunk 11]
```diff
@@ -272,36 +282,14 @@ nextApp.prepare().then(() => {
...
-      const backing = getDependenciesAvailableForBacking(recommendations);
-      return res.status(200).send(backing);
-    } catch (err) {
-      console.error(err);
-      return res.status(400).send('Unable to fetch file');
-    }
-  });
-
-  server.get('/:id/profile/backing.json', async (req, res) => {
+  server.get('/:id/:type/backing.json', async (req, res) => {
```
Comment (znarf) [inline src/server/index.js]: What is `:type` used for?

File: src/pages/monthly-plan.js [hunk 12]
```diff
@@ -62,6 +52,24 @@ const suggestedAmounts = [
   },
 ];
 
+const getProfileData = (id, accessToken, excludedRepos) => {
```
Comment (znarf) [inline src/pages/monthly-plan.js]: This looks duplicated with the code in `fetch.js`

File: src/pages/monthly-plan.js [hunk 13]
```diff
@@ -62,6 +52,24 @@ const suggestedAmounts = [
...
+  }
+  const searchParams = new URLSearchParams(params);
+  return process.env.IS_CLIENT
+    ? fetchJson(`/data/getProfileData?${searchParams}`)
+    : import('../lib/data').then(m =>
+        m.getProfileData(id, accessToken, { excludedRepos }),
+      );
+};
+
+const getFilesData = sessionFiles =>
```
Comment (znarf) [inline src/pages/monthly-plan.js]: Why do we have that there and not in `fetch.js` ?

File: src/pages/monthly-plan.js (hunk 12 above)
Comment (flickz) [reply src/pages/monthly-plan.js]: Oops that should not be there again, I thought I removed it.

File: src/pages/monthly-plan.js (hunk 13 above)
Comment (flickz) [reply src/pages/monthly-plan.js]: I initially moved it to fetch.js  as an abstraction because it was  both in profile and monthly plan and created a single function `getData` to use for both file and profile type. But you asked me to remove it and revert back. I forgot to delete it from fetch.js after removing `getData`.

File: src/server/index.js (hunk 11 above)
Comment (flickz) [reply src/server/index.js]: We needed it before but since reverting back to saving both private and public files on s3, we don't need it at this point but I'm skeptical about removing it because we might need it in the future and it will be difficult to add it again when we subscriptions depending on the endpoint for dispatch.

Comment (flickz) [reply src/server/index.js]: @znarf Also if I remove it here, we'll have a duplicate endpoint with https://github.com/opencollective/backyourstack/blob/feat%2Fselect-repos/src/server/index.js#L255

File: src/lib/s3.js [hunk 14]
```diff
@@ -71,27 +72,30 @@ export const getObjectList = id => {
...
 };
 
-export const getProfileSavedData = async id => {
+export const getObjectsMetadata = async id => {
+  const params = { Bucket, Key: `${id}/dependencies.json` };
+  const { Body } = await s3.getObject(params).promise();
+  return JSON.parse(Body.toString('utf-8'));
+};
+
+export const getData = async id => {
```
Comment (znarf) [inline src/lib/s3.js]: This is not clear what it's doing and looks out of place in the s3 lib. s3 should not import the data lib.

I would refactor that out and clarify the `data` variables.

**Comment by znarf** [comment]
Closing the PR, too many iterations, too many comments.

When you're ready, please open a new one, **on a new a branch**, with a single commit.
